File: Sources/PowerpointViewer/XMLExtensions.swift
import Foundation

/// Namespace-agnostic traversal helpers for `XMLElement`.
///
/// OOXML documents are heavily namespaced and different producers use different
/// prefixes, so every query matches on the element's *local* name.
///
/// Implementation note: these walk the tree manually rather than using
/// `nodes(forXPath:)`. Foundation's XPath evaluation has been observed escaping
/// the context node on some documents (a `.//*` query returning matches from the
/// whole document), which silently corrupted color resolution — a background's
/// `solidFill` would pick up an unrelated `srgbClr` from elsewhere in the slide.
extension XMLElement {
    /// All descendants (any depth) with the given local name, in document order.
    func descendants(localName: String) -> [XMLElement] {
        var result: [XMLElement] = []
        func walk(_ element: XMLElement) {
            for case let child as XMLElement in element.children ?? [] {
                if child.localName == localName { result.append(child) }
                walk(child)
            }
        }
        walk(self)
        return result
    }

    /// First descendant (any depth, document order) with the given local name.
    func firstDescendant(localName: String) -> XMLElement? {
        for case let child as XMLElement in children ?? [] {
            if child.localName == localName { return child }
            if let found = child.firstDescendant(localName: localName) { return found }
        }
        return nil
    }

    /// First *direct child* with the given local name.
    func firstChild(localName: String) -> XMLElement? {
        for case let child as XMLElement in children ?? [] {
            if child.localName == localName { return child }
        }
        return nil
    }
}

File: Windows/SlideViewer/Parsing/XmlExtensions.cs
using System.Xml.Linq;

namespace SlideViewer.Parsing;

/// <summary>
/// Namespace-agnostic traversal helpers. OOXML uses many namespace prefixes
/// (p:, a:, r:) that vary by producer, so everything matches on local name.
/// Traversal is explicit (never XPath) — the macOS build hit a Foundation XPath
/// bug where ".//" escaped the context node and corrupted color resolution.
/// </summary>
public static class XmlExtensions
{
    /// <summary>First direct child with the given local name.</summary>
    public static XElement? Child(this XElement? element, string localName) =>
        element?.Elements().FirstOrDefault(e => e.Name.LocalName == localName);

    /// <summary>Direct children with the given local name, in document order.</summary>
    public static IEnumerable<XElement> Children(this XElement? element, string localName) =>
        element?.Elements().Where(e => e.Name.LocalName == localName) ?? Enumerable.Empty<XElement>();

    /// <summary>First descendant (any depth, document order) with the local name.</summary>
    public static XElement? Descendant(this XElement? element, string localName)
    {
        if (element == null) return null;
        foreach (var child in element.Elements())
        {
            if (child.Name.LocalName == localName) return child;
            var found = child.Descendant(localName);
            if (found != null) return found;
        }
        return null;
    }

    /// <summary>All descendants with the given local name, in document order.</summary>
    public static IEnumerable<XElement> DescendantsNamed(this XElement? element, string localName)
    {
        if (element == null) yield break;
        foreach (var child in element.Elements())
        {
            if (child.Name.LocalName == localName) yield return child;
            foreach (var nested in child.DescendantsNamed(localName)) yield return nested;
        }
    }

    /// <summary>Attribute by local name, preferring the unprefixed one so a plain
    /// `id` wins over `r:id`.</summary>
    public static string? Attr(this XElement? element, string localName)
    {
        if (element == null) return null;
        var direct = element.Attribute(localName);
        if (direct != null) return direct.Value;
        return element.Attributes()
            .FirstOrDefault(a => a.Name.LocalName == localName)?.Value;
    }

    /// <summary>Relationship attribute from the r: namespace (r:id, r:embed).</summary>
    public static string? RelAttr(this XElement? element, string localName) =>
        element?.Attributes()
            .FirstOrDefault(a => a.Name.LocalName == localName &&
                                 a.Name.NamespaceName.Contains("relationships"))?.Value;

    public static double? AttrDouble(this XElement? element, string localName)
    {
        var s = element.Attr(localName);
        return double.TryParse(s, System.Globalization.NumberStyles.Any,
            System.Globalization.CultureInfo.InvariantCulture, out var v) ? v : null;
    }

    public static int? AttrInt(this XElement? element, string localName)
    {
        var s = element.Attr(localName);
        return int.TryParse(s, out var v) ? v : null;
    }
}
